{"cells":[{"cell_type":"code","source":["# Just run the commands below if you are using colab. Otherwise, don't run.\n","\n","from google.colab import drive\n","drive.mount('/content/drive', force_remount=True)\n","\n","import os\n","os.chdir('drive/MyDrive/mestrado/pipeline/scripts')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"BVOc6neSGPaV","executionInfo":{"status":"ok","timestamp":1693863251815,"user_tz":180,"elapsed":2217,"user":{"displayName":"HELMUT PATROCINIO","userId":"01392365920447132820"}},"outputId":"6890675a-745b-4373-cf30-5ac9713ab3f5"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"]}]},{"cell_type":"code","source":["def package_installed(package):\n","  try:\n","    import package\n","    print(f\"{package} installed\")\n","  except:\n","    os.system(\"pip install \"+ package)\n","\n","package_installed(\"xlsxwriter\")\n","package_installed(\"biopython\")"],"metadata":{"id":"9la7v-MVGMiu"},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"metadata":{"id":"15Uvi22CEp_k"},"outputs":[],"source":["import pandas as pd\n","from xlsxwriter.workbook import Workbook\n","from Bio import pairwise2 as pw\n","from Bio.Align import substitution_matrices as sm\n","import glob\n","import regex as re\n","import os\n","import itertools"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"7mkO95ODEp_q"},"outputs":[],"source":["def create_excel_file(protein):\n","    global worksheet\n","    \"\"\" Create workbook and worksheet objects for a protein of interest\n","\n","    Args:\n","        protein ('_io.TextIOWrapper'): File containing the sequences of the protein of interest\n","    Returns:\n","        Workbook and worksheet objects\n","    \"\"\"\n","    path = os.path.join(\"..\", \"IdentityAnalysis\")\n","    isExist = os.path.exists(path)\n","    if not isExist:\n","    # Create a new directory because it does not exist\n","        os.makedirs(path)\n","        print(\"The new directory is created!\")\n","\n","    protein_name = re.search(r'.*(?=\\.fasta)', protein.name).group()\n","    dir_name = os.path.join(\"..\", \"IdentityAnalysis\", f\"{protein_name}.xlsx\")\n","    workbook = Workbook(dir_name)\n","    worksheet = workbook.add_worksheet()\n","\n","    return workbook, worksheet\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"eS65at32Ep_r"},"outputs":[],"source":["def style_text(workbook):\n","    global style1, style2\n","    \"\"\"\n","    Defines a formatting style to highlight the identity between two amino acid residues.\n","    Two equal residues will be formatted with 'vivid blue' (#058ED9) color, bold and font size 14.\n","    Two different residues will be formatted with black color and font size 12.\n","\n","    Args:\n","        workbook: xlsxwriter.workbook.Workbook type object\n","\n","    Returns:\n","        Objects xlsxwriter.format.Format: style1 and style2\n","    \"\"\"\n","    style1 = workbook.add_format({'color': '#058ED9', 'bold': True, 'font_size': 14, 'font':'arial'})\n","    style2 = workbook.add_format({'color': 'black', 'font_size': 12, 'font':'arial'})\n","    return style1, style2"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"dDRbIdEdEp_s"},"outputs":[],"source":["def get_nonamers_and_headers(file):\n","    \"\"\"Get amino acid sequences (nonamers) and headers from a fasta file\n","\n","    Args:\n","        file (_io.TextIOWrapper): Fasta file containing sequences of the protein(s) of interest\n","\n","    Returns:\n","        Lists containing the nonamers and headers of a fasta file\n","    \"\"\"\n","    index = 1\n","\n","    headers = []\n","    nonamers = []\n","    for line in file:\n","        line = line.rstrip(\"\\n\")\n","        \"\"\"\n","        When index is odd, the corresponding line of text must be added to the header list.\n","        Otherwise, it is added to the nonamers list.\n","        \"\"\"\n","        if index % 2 == 1:\n","            # The first character of the header is the '>', so the content is stored from position 1 (second character)\n","            headers.append(line[1:])\n","        else:\n","            nonamers.append(line)\n","        index += 1\n","\n","    return nonamers, headers\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"gnfB4n6zEp_t"},"outputs":[],"source":["def select_rule(identity):\n","    \"\"\"\n","    Identifies rule for which the pair of nonamers was selected.\n","\n","    General rule: An identical residue in P5\n","    Rule 1: All residues at positions P2, P3 and P8 are identical.\n","    Rule 2: Two identical residues in P2, P3 or P8 and at least one in P4, P6 or P7.\n","    Rule 3: All residues in positions P4, P6 and P7 are identical and at least one in P2, P3 ou P8.\n","\n","    Args:\n","        identity (int): identity between two nonamers\n","    Returns:\n","        identity (str): rule for which pair of nonamers was selected\n","    \"\"\"\n","    rules = {30:'Rule 1', 33:'Rule 1', 36:'Rule 1', 39:'Rule 1',\n","             28:'Rule 2', 31:'Rule 2', 34:'Rule 2', 29:'Rule 3'}\n","    rule = rules[identity]\n","    return rule"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"9dBXCShUEp_u"},"outputs":[],"source":["def identity_TCR_contacts(nonamer_1, nonamer_2):\n","    \"\"\"\n","    Checks if there is identity between two residues located in the same position.\n","    The residue is formatted with style 1 if identical, or style 2 if not identical.\n","\n","    Args:\n","        nonamer_1 (str): nonamer from the protein of interesting 1\n","        nonamer_2 (str): nonamer from the protein of interesting 2\n","        style1 (xlsxwriter.format.Format): Formatting style for identical residues ('vivid blue' color, bold and font size 14)\n","        style2 (xlsxwriter.format.Format): Formatting style for different residues (black color and font size 12)\n","\n","    Returns:\n","        Lists containing the stylized residues of nonamers from the proteins of interest 1 (formatting_style_1) and 2 (formatting_style_2).\n","        It also returns the value of the identity between the two nonamers.\n","    \"\"\"\n","    formatting_style_1 = []\n","    formatting_style_2 = []\n","    identity = 0\n","\n","    # The identity between residues is scored according to the relevance of their position for recognition by the TCR\n","\n","    positions = {'P1': 0.0, 'P2': 5, 'P3': 5, 'P4': 3, 'P5': 15, 'P6': 3, 'P7': 3, 'P8': 5, 'P9': 0.0}\n","\n","    values = list(positions.values())\n","    position = 0\n","\n","    # Nonamer_1 and nonamer_2 are strings, so the zip function creates tuples containing characters located in the same position.\n","    # Ex.: zip('ABC', 'DEF') -> (('A','D'), ('B','E'), ('C','F'))\n","\n","    for aa_1, aa_2 in zip(nonamer_1, nonamer_2):\n","        if aa_1 == aa_2:\n","            formatting_style_2.extend((style1, aa_2))\n","            formatting_style_1.extend((style1, aa_1))\n","            identity += values[position]\n","\n","        else:\n","            formatting_style_2.extend((style2, aa_2))\n","            formatting_style_1.extend((style2, aa_1))\n","\n","        position += 1\n","\n","    return formatting_style_1, formatting_style_2, identity\n"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"_MlpDK65Ep_v","executionInfo":{"status":"ok","timestamp":1693863208756,"user_tz":180,"elapsed":54,"user":{"displayName":"HELMUT PATROCINIO","userId":"01392365920447132820"}},"outputId":"fcae0469-af25-4a36-d31d-3e3c95b700f2"},"outputs":[{"output_type":"execute_result","data":{"text/plain":["True"]},"metadata":{},"execution_count":19}],"source":["pares_residuos = dict(enumerate(zip('TTGARVELS', 'AEGLRVLLA'), start=1))\n","aa_1, aa_2 = pares_residuos[5]\n","aa_1 == aa_2"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"zKNU_xW0Ep_x"},"outputs":[],"source":["def hla_analysis(haplotype_list, hla_1, hla_2):\n","    \"\"\" Checks if the two alleles are the same or if they are present in the same haplotype\n","\n","    Args:\n","        haplotype_list (list): list containing the HLA-DQ and HLA-DR pairs that are present in the same haplotype haplotype\n","        hla_1 (str): HLA for which the nonamer of the protein of interest 1 was predicted\n","        hla_2 (str): HLA for which the nonamer of the protein of interest 2 was predicted\n","\n","    Returns:\n","        (bool): Returns True if the two alleles are the same or if they are present in the same haplotype. Otherwise, it returns False.\n","    \"\"\"\n","    same_alleles = (hla_1 == hla_2)\n","\n","    if same_alleles:\n","        return True\n","\n","    for haplotype in haplotype_list:\n","        allele_1_present = hla_1 in haplotype\n","        allele_2_present = hla_2 in haplotype\n","\n","        if allele_1_present and allele_2_present:\n","            return True\n","\n","    return False\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"RXQRXjyjEp_y"},"outputs":[],"source":["def selected_haplotypes():\n","    \"\"\"\n","    Load file containing the HLA-DQ and HLA-DR pairs that are present in haplotypes\n","\n","    Returns:\n","        haplotype_list (list): list containing the HLA-DQ and HLA-DR pairs that form haplotypes\n","    \"\"\"\n","    dir_name = os.path.join(\"..\", \"..\", \"files\", \"selected_haplotypes.txt\")\n","    file_haplotypes = open(dir_name)\n","    haplotype_list = file_haplotypes.read().split('\\n')\n","    file_haplotypes.close()\n","    return haplotype_list\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"BTeV0z-8Ep_z"},"outputs":[],"source":["def extract_headers_data(headers):\n","    \"\"\"\n","    Find the HLA present in the nonamer header\n","\n","    Args:\n","        headers (str): Header in the following format: 'protein name (HLA allele)'\n","        Ex.: CNP (HLA-DRB1*04:01)\n","\n","    Returns:\n","        hlas(list): List of HLA alleles\n","        new_headers(list): List of new headers\n","    \"\"\"\n","\n","    # The code below finds the HLA in the header. If it is HLA-DQ, the '/' is replaced by '-'\n","    # Ex.: CNP (HLA-DRB1*04:01) -> DRB1*04:01\n","    # Ex.: MAG (HLA-DQA1*05:01/DQB1*02:01) -> DQA1*05:01-DQB1*02:01\n","\n","    pattern_hla_find = re.compile(r'(?<=\\(HLA-).*\\d{2}:\\d{2}(?=\\))')\n","    hlas = [re.search(pattern_hla_find, headers).group().replace(\"/\", \"-\")\n","            if re.search(pattern_hla_find, headers)\n","            else ''\n","            for headers in headers]\n","\n","    # The code below finds the HLA in the header and replaces it with nothing\n","    # Ex.: 'CNP (HLA-DRB1*04:01)' > CNP\n","    # Ex.: 'BHRF1 (HLA-DRB1*12:01) (B95-8 - GD1)' > 'BHRF1  (B95-8 - GD1)'\n","\n","    pattern_hla_sub = re.compile(r'\\(HLA-.*\\d{2}:\\d{2}\\)')\n","    new_headers = [re.sub(pattern_hla_sub, '', headers) for headers in headers]\n","\n","    return hlas, new_headers\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Jwe3lfuIEp_0"},"outputs":[],"source":["def combines_alleles_and_nonamers(headers_1, headers_2, nonamers_1, nonamers_2):\n","    \"\"\" Receives headers and nonamers from the proteins being analyzed.\n","        Generates all possible combinations of pairs between the nonamers of the two proteins.\n","        Generates identifiers for each nonamers and the respective HLA for which it was predicted.\n","\n","    Args:\n","        headers_1 (list): List containing the headers of the protein of interest 1.\n","        headers_2 (list): List containing the headers of the protein of interest 2.\n","        nonamers_1 (list): List containing the nonamers of the protein of interest 1.\n","        nonamers_2 (list): List containing the nonamers of the protein of interest 2.\n","\n","    Returns:\n","        hlas_nonamers_combination (list): List containing all possible combinations between the nonamers of the proteins being analyzed.\\n\n","        See the example below to understand what this variable represents:\n","\n","            Nonamer from protein APLP1 presented by HLA-DRB1*01:01\n","                >APLP1 (HLA-DRB1*01:01)\n","                LLLLRAQPA\n","\n","            Five nonamers from Epstein-Barr Virus\n","                >BHRF1 (HLA-DRB1*12:01) (B95-8 - GD1)\n","                FLAGLTLSL\\n\n","                >BHRF1 (HLA-DRB1*11:01) (B95-8 - GD1)\n","                LLALCIRDS\\n\n","                >BHRF1 (HLA-DRB1*11:01) (B95-8 - GD1)\n","                YLFISRGRH\\n\n","                >BHRF1 (HLA-DRB4*01:01) (B95-8 - GD1)\n","                YHVLLEEII\\n\n","                >BHRF1 (HLA-DRB4*01:01) (B95-8 - GD1)\n","                TLSLLVICS\\n\n","\n","            The HLA DRB1*01:01 of APLP1 is combined with all five HLAs from EBV\\n\n","            HLA_pairs = [('DRB1*01:01', 'DRB1*12:01'),\\n\n","                         ('DRB1*01:01', 'DRB1*11:01'),\\n\n","                         ('DRB1*01:01', 'DRB1*11:01'),\\n\n","                         ('DRB1*01:01', 'DRB4*01:01'),\\n\n","                         ('DRB1*01:01', 'DRB4*01:01')]\n","\n","            The nonamer LLLLRAQPA of APLP1 is combined with all five nonamers from EBV\\n\n","            nonamers_pairs = [('LLLLRAQPA', 'FLAGLTLSL'),\\n\n","                             ('LLLLRAQPA', 'LLALCIRDS'),\\n\n","                             ('LLLLRAQPA', 'YLFISRGRH'),\\n\n","                             ('LLLLRAQPA', 'YHVLLEEII'),\\n\n","                             ('LLLLRAQPA', 'TLSLLVICS')]\n","\n","\n","            Each tuple in the list contains the nonamers and HLAs for which they were predicted.\n","            These nonamers will be evaluated for the identity of their residues.\\n\n","            hlas_nonamers_combination = [('DRB1*01:01', 'DRB1*12:01', 'LLLLRAQPA', 'FLAGLTLSL'),\\n\n","                                         ('DRB1*01:01', 'DRB1*11:01', 'LLLLRAQPA', 'LLALCIRDS'),\\n\n","                                         ('DRB1*01:01', 'DRB1*11:01', 'LLLLRAQPA', 'YLFISRGRH'),\\n\n","                                         ('DRB1*01:01', 'DRB4*01:01', 'LLLLRAQPA', 'YHVLLEEII'),\\n\n","                                         ('DRB1*01:01', 'DRB4*01:01', 'LLLLRAQPA', 'TLSLLVICS')]\n","\n","        identifier_1 (dict): Dictionary containing the header keys and the values that identify them.\n","                                Ex.: Header -> CNP (HLA-DRB1*04:01)\n","                                     Nonamer -> FRKMSSSGA\n","                                     HLA -> DRB1*04:01\n","                                     Identifier -> DRB1*04:01-FRKMSSSGA\n","        identifier_2 (dict): Dictionary containing the header keys and the values that identify them.\n","    \"\"\"\n","\n","    HLAs_1, new_headers_1 = extract_headers_data(headers_1)\n","    HLAs_2, new_headers_2 = extract_headers_data(headers_2)\n","\n","    HLA_pairs = itertools.product(HLAs_1, HLAs_2)\n","    nonamers_pairs = itertools.product(nonamers_1, nonamers_2)\n","    hlas_nonamers_combination = [\n","        alleles + nonamers for alleles, nonamers in (zip(HLA_pairs, nonamers_pairs))]\n","\n","    identifier_1 = [\"-\".join([h, n])\n","                        for h, n in [*zip(HLAs_1, nonamers_1)]]\n","    identifier_1 = dict(zip(identifier_1, new_headers_1))\n","\n","    identifier_2 = [\"-\".join([h, n])\n","                       for h, n in [*zip(HLAs_2, nonamers_2)]]\n","    identifier_2 = dict(zip(identifier_2, new_headers_2))\n","\n","    return hlas_nonamers_combination, identifier_1, identifier_2\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"4q_jVzTrEp_0"},"outputs":[],"source":["def calculate_similarity_score(nonamer_1, nonamer_2):\n","    \"\"\"\n","    Calculates the blosum62 similarity score between the central region of the nonamers\n","\n","    Args:\n","        nonamer_1 (str): Nonamer from protein of interest 1\n","        nonamer_2 (str): Nonamer from protein of interest 2\n","\n","    Returns:\n","        similarity_score (int): Similarity score of the central region of the nonamers (P2 to P8)\n","    \"\"\"\n","    blosum62 = sm.load(\"BLOSUM62\")\n","    aln = pw.align.globalds(\n","        nonamer_1[1:8], nonamer_2[1:8], blosum62, -5, -1)[0]\n","    result = pw.format_alignment(*aln)\n","    similarity_score = int(re.search('\\d+', result).group())\n","    return similarity_score\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"DmIzquGHEp_0"},"outputs":[],"source":["def write_line_in_excel(line, formatting_style_1, formatting_style_2, identifier_1, identifier_2, HLA_1,\n","                        HLA_2, nonamer_1, nonamer_2, similarity_score, selection_rule):\n","    \"\"\"\n","    Write a line in the worksheet object\n","\n","    Args:\n","        worksheet (worksheet): Worksheet object in which the lines with the result of the identity analysis between the nonamers will be written\n","        line (int): Number that indicates the position of the line that should receive the results\n","        formatting_style_1 (list): List containing the formatted characters of nonamers\n","        formatting_style_2 (list): List containing the formatted characters of nonamers\n","        identifier_1 (dict): Dictionary containing the headers in the keys and the values that identify them.\n","        identifier_2 (dict): Dictionary containing the headers in the keys and the values that identify them.\n","        HLA_1 (str): HLA allele\n","        HLA_2 (str): HLA allele\n","        nonamer_1 (str): Nonamers from protein of interest 1\n","        nonamer_2 (str): Nonamers from protein of interest 2\n","        similarity_score (int): Similarity score between nonamers\n","    \"\"\"\n","    identifier_1 = identifier_1[f'{HLA_1}-{nonamer_1}']\n","    identifier_2 = identifier_2[f'{HLA_2}-{nonamer_2}']\n","\n","    # Defines the texts (the column names) of the first eight cells in the first row.\n","    columns = ['Protein_1', 'HLA_1', 'Nonamer_1', 'Nonamer_2', 'HLA_2',\n","               'Protein_2', 'Similarity Score BLOSUM62', 'Selection Rule']\n","\n","    # Write column names from cell A1\n","    worksheet.write_row('A1', columns, style2)\n","\n","    # Write the results of each analysis.\n","    worksheet.write(line, 0, identifier_1, style2)\n","    worksheet.write(line, 1, HLA_1, style2)\n","    worksheet.write_rich_string(line, 2, *formatting_style_1)\n","    worksheet.write_rich_string(line, 3, *formatting_style_2)\n","    worksheet.write(line, 4, HLA_2, style2)\n","    worksheet.write(line, 5, identifier_2, style2)\n","    worksheet.write(line, 6, similarity_score, style2)\n","    worksheet.write(line, 7, selection_rule, style2)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"yjt5p7whEp_1"},"outputs":[],"source":["def nonamer_identity_analysis(file_1, file_2):\n","    \"\"\"\n","    Function responsible for calling the other functions necessary for the analysis of identity between nonamers.\n","\n","    Args:\n","        file_1 (_io.TextIOWrapper): File containing the nonamers of the protein of interest 1\n","        file_2 (_io.TextIOWrapper): File containing the nonamers of the protein of interest 2\n","        worksheet (worksheet): Object on which the analysis results will be written\n","        style1 (format): Formatting style for identical residues\n","        style2 (format): Formatting style for different  residues\n","    \"\"\"\n","    line_excel = 1\n","    nonamers_1, headers_1 = get_nonamers_and_headers(file_1)\n","    nonamers_2, headers_2 = get_nonamers_and_headers(file_2)\n","    hlas_nonamers_combination, identifier_1, identifier_2 = combines_alleles_and_nonamers(\n","        headers_1, headers_2, nonamers_1, nonamers_2)\n","    haplotype_list = selected_haplotypes()\n","\n","    for hla_1, hla_2, nonamer_1, nonamer_2 in hlas_nonamers_combination:\n","        equal_hlas_or_haplotype = hla_analysis(haplotype_list, hla_1, hla_2)\n","\n","        if equal_hlas_or_haplotype:\n","            formatting_style_1, formatting_style_2, identity = identity_TCR_contacts(nonamer_1, nonamer_2)\n","        else:\n","            continue\n","\n","        if identity >= 28:\n","            selection_rule = select_rule(identity)\n","            similarity_score = calculate_similarity_score(nonamer_1, nonamer_2)\n","            write_line_in_excel(line_excel, formatting_style_1, formatting_style_2, identifier_1, identifier_2, hla_1,\n","                                hla_2, nonamer_1, nonamer_2, similarity_score, selection_rule)\n","            line_excel += 1"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"bhp_TS_iEp_1"},"outputs":[],"source":["def main(peptides_1, peptides_2):\n","    \"\"\"\n","    Function responsible for reading the files with the nonamers\n","    and running the create_excel_file, style_text and analyze_identity_nonamers functions.\n","\n","    Args:\n","        peptides_1 (str): Name of the fasta file with the nonamers of the protein of interest 1.\n","        peptides_2 (str): Name of the fasta file with the nonamers of the protein of interest 2.\n","    \"\"\"\n","    file_1 = open(peptides_1)\n","    file_2 = open(peptides_2)\n","    workbook, worksheet = create_excel_file(file_1)\n","    style1, style2 = style_text(workbook)\n","\n","    nonamer_identity_analysis(file_1, file_2)\n","\n","    workbook.close()\n","    file_1.close()\n","    file_2.close()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"oTKic4kzEp_1"},"outputs":[],"source":["try:\n","    path = os.path.join(\"..\", \"results\", \"nonamers_fastas\")\n","    os.chdir(path)\n","except:\n","    pass"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"3azy4BeaEp_1","executionInfo":{"status":"ok","timestamp":1693863315486,"user_tz":180,"elapsed":22723,"user":{"displayName":"HELMUT PATROCINIO","userId":"01392365920447132820"}},"outputId":"d2e57ab3-4c20-4fa0-a806-6327e1c65ca9"},"outputs":[{"output_type":"stream","name":"stdout","text":["Periaxin_sn_core.fasta\n","P0 protein_sn_core.fasta\n","CNP_sn_core.fasta\n","PLP_sn_core.fasta\n","MBP_sn_core.fasta\n","GlialCam_sn_core.fasta\n","MAG_sn_core.fasta\n","P2 protein_sn_core.fasta\n","APLP1_sn_core.fasta\n","MOG_sn_core.fasta\n","MOBP_sn_core.fasta\n"]}],"source":["for proteina in glob.glob(r\"*_sn_core.fasta\"):\n","    print(proteina)\n","    main(proteina, 'ebv_core.fasta')"]}],"metadata":{"interpreter":{"hash":"d66488630c501ba952a1dd24d15309edbda65c7d8ac53ee663f5bb4f81535211"},"kernelspec":{"display_name":"Python 3.10.4 64-bit","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.4"},"orig_nbformat":4,"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":0}